tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/TargetKeyTest.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.activemq.artemis.core.server.balancing.policies.PolicyFactoryResolver;
+import org.apache.activemq.artemis.core.server.balancing.targets.Target;
+import org.apache.activemq.artemis.core.server.balancing.targets.TargetKey;
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+@RunWith(Parameterized.class)
+public class TargetKeyTest extends BalancingTestBase {
+
+   private static final String MOCK_POLICY_NAME = "MOCK_POLICY";
+
+   @Parameterized.Parameters(name = "protocol: {0}")
+   public static Collection<Object[]> data() {
+      Collection<Object[]> data = new ArrayList<>();
+
+      for (String protocol : Arrays.asList(new String[] {AMQP_PROTOCOL, CORE_PROTOCOL, OPENWIRE_PROTOCOL})) {
+         data.add(new Object[] {protocol});
+      }
+
+      return data;
+   }
+
+
+   private final String protocol;
+
+   private final List<String> keys = new ArrayList<>();
+
+
+   public TargetKeyTest(String protocol) {
+      this.protocol = protocol;
+   }
+
+   @Before
+   public void setup() throws Exception {
+      PolicyFactoryResolver.getInstance().registerPolicyFactory(
+         new PolicyFactory() {
+            @Override
+            public String[] getSupportedPolicies() {
+               return new String[] {MOCK_POLICY_NAME};
+            }
+
+            @Override
+            public Policy createPolicy(String policyName) {
+               return new FirstElementPolicy(MOCK_POLICY_NAME) {
+                  @Override
+                  public Target selectTarget(List<Target> targets, String key) {
+                     keys.add(key);
+                     return super.selectTarget(targets, key);
+                  }
+               };
+            }
+         });
+   }
+
+   @Test
+   public void testClientIDKey() throws Exception {
+      setupLiveServerWithDiscovery(0, GROUP_ADDRESS, GROUP_PORT, true, true, false);
+      setupBalancerServerWithDiscovery(0, TargetKey.CLIENT_ID, MOCK_POLICY_NAME, null, true, null, 1);
+      startServers(0);
+
+      ConnectionFactory connectionFactory = createFactory(protocol, false, TransportConstants.DEFAULT_HOST,
+         TransportConstants.DEFAULT_PORT + 0, "test", null, null);
+
+      keys.clear();
+
+      try (Connection connection = connectionFactory.createConnection()) {
+         connection.start();
+      }
+
+      Assert.assertEquals(1, keys.size());
+      Assert.assertEquals("test", keys.get(0));
+   }
+
+   @Test
+   public void testSNIHostKey() throws Exception {
+      final String localHostname = "localhost.localdomain";
+
+      assumeLocalHostname(localHostname);
```

Comment:
Personally I'd still prefer a test that ran everywhere connecting to localhost (with non-localhost SNI with Core client possible), over a test that needs checking if it ran. A unit test could probably also be used to ensure the connection is passed to the validator, and that it sees non-"localhost" values pass through from the connection.

Reply:
Now the test is executed at least with CORE protocol

Reply:
Not quite what I was suggesting but hopefully does the trick with the fallbacks :)

(I was suggesting, have the core client connect to "localhost" but override the SNI to something else - like "localhost.localdomain" as the cert already matches that too)